### ISSUE TITLE
Can you compile a Linux release?

### ISSUE THREAD

**Issue by GAZ082** [Macad3D/Macad3D#50]
Please? :D

**Comment by dipts** [comment]
Sorry, there is currently no linux port. Although basically feasible, there is a lot of work behind it. Since I don't use linux as desktop I won't make the port myself. But I'm happy to support anyone who wants to try.

**Comment by app4soft** [comment]
> Sorry, there is currently no linux port.

As .NET 6 available for Linux and Mac (and there is Mono Runtime too), what else needs to be ported?

Has Macad3D any Windows-specific dependency?

**Comment by dipts** [comment]
The core functionality - the model, interactive tools, exchange, etc. - have no dependency on Windows. As far as I can see, there are two major areas that need to be ported:

- The UI is based on WPF, which is currently only available on Windows. This would have to be changed to another framework. Avalonia would be a good choice, as it is also XAML-based and very similar to WPF. With this, the majority of XAMLs and code-behind should be able to be transferred.

- There is currently no Linux compiler that compiles C++/CLI. The wrapping for OCCT (and Macad's own CPP components) must be converted to P/Invoke. The wrapper generation is already automated, so only the generator has to be adapted, not the consuming code.

The rest should be peanuts. Some OS dependent functions have to be encapsulated and implemented for Linux, the build system needs to be enhanced, such things. 

All in all doable, just a lot of work. If this can be distributed on a few shoulders it becomes realistic.

**Comment by cactushydrocodone** [comment]
Ive got Macad3D to start (show me the start selection screen, show gui) with the following steps.
Keep in mind i reconstructed this out of my head after trying several things, some steps might not be necessary.

As an example, i use `/wine/prefix` as wine prefix path here. 

1. install wine-staging and winetricks
2. Run winetricks the following way: `WINEPREFIX=/wine/prefix winetricks` (This prefix will be created with winetricks)
2.1. Set windows version to windows 10
2.2. Install dxvk so wpf support is given
2.3. install some basic windows fonts like arial, tahoma, etc
3. Download and install the 6.0.12 dotnet runtime (windowsdesktop-runtime-6.0.12-win-x64.exe) using `WINEPREFIX=/wine/prefix ~/Downloads/windowsdesktop-runtime-6.0.12-win-x64.exe` (Mark the file as executable first using `chmod +x`)
4. Download the Macad3D installer from the releases page (dont forget, mark as executable)
5. Start the Macad3D installer using `WINEPREFIX=/wine/prefix ~/Downloads/Macad3D_2.3_Setup.exe`
6. Click it through, do NOT run it at the end, e.g. untick the checkbox.
7. Now, for whatever reason, wine defaults to searching for libraries on the wine host instead in the prefix. To make macad actually run, use the following command: `WINEPREFIX=/wine/prefix DOTNET_ROOT="C:\Program Files\dotnet" /wine/prefix/drive_c/Program\ Files/Macad3D/Macad.exe`

This will make the program start, but its not usable yet.

**Comment by cactushydrocodone** [comment]
![image](https://github.com/Macad3D/Macad3D/assets/12892508/7a4600f1-a1cc-433f-9175-20366328e897)
This is what i mean by "weird black windows"

**Comment by dipts** [comment]
Interesting experiment, that doesn't look so bad indeed.

The "weird black window" is a transparent window that holds UI over the viewport. This actually arranges itself automatically over the entire viewport area and should not be movable. The black frame around the about window is also transparent and contains the shadow effect.

The coordinate deviation may be due to the fact that under Windows an offset must be included if you want to paint the window over the entire window area (and not only within the client area). This calculation depends on how large the OS Window Manager creates the window frame. This explains the improvement in fullscreen mode. Maybe a compatibility mode could be added here, if Wine is detected, which restricts the rendering to the client area or uses different window settings.

Scripting might work, it just needs the .Net runtime.

**Comment by xyont** [comment]
indeed, Linux versions will attract large user and programmer also, maybe an improvement and plugins.

**Comment by farleyrunkel** [comment]
hi，how about a cpp based macad for cross-platform?

**Comment by dipts** [comment]
With version 4.0 I have introduced some fixes to increase the compatibility with Wine. I have tested on Ubuntu 23, works fine so far. 

I have also created a [documentation page](https://macad3d.net/userguide/go/?guid=46793f6d-da23-48ea-913a-186727062a80) with hints on how to install under Wine. If any of you have any tips or hints, you can edit this page directly in GitHub and add to it.

Since nobody has volunteered to replace the GUI so far (the programming language is not the problem...) I hope to be able to make the app available under Linux in this way.

Feedback is welcome.

**Comment by dipts** [comment]
This seems to be a regression of Wine 10, which is fixed in the current staging branch. I tested with Ubuntu 24 and Wine 10.19, it is not crashing anymore.

**Comment by Raketenphysik** [comment]
Thanks for this great software. Seems to be the only sane GUI based FOSS CAD imho!
Just wanted to let you know that Macad runs like a charm via Bottles. Just check "Discrete Graphics" via settings.
